core/lib/config/src/configs/da_client/mod.rs

```diff
@@ -0,0 +1,19 @@
+use serde::Deserialize;
+
+use crate::{AvailConfig, ObjectStoreConfig};
+
+pub mod avail;
+
+#[derive(Debug, Clone, PartialEq, Deserialize)]
+pub struct DAClientConfig {
+    #[serde(flatten)]
+    pub client: DAClient,
```

Comment:
[This](https://github.com/softprops/envy/issues/26#issuecomment-1072980219) seems to be a better approach to implement this functionality compared to custom deserializer.

We only rely on `serde` as an implementation detail; in the code outside of the config crate we rely on either `FromEnv` trait or the protobuf codegen.

So what you can do is to remove `Deserialize` from this structure, and then use `envy` to deserialize `DAClient` inside of the `FromEnv` implementation. Please leave a comment explaining why this structure doesn't implement `Deserialize` though.

Reply:
Then you can implement `flatten` manually, e.g. (pseudocode)

```rust
fn from_env() -> anyhow::Result<Self> {
  let client_tag = std::env::var("DA_CLIENT")?;
  let res = match client_tag {
    Self::AVAIL_CLIENT_CONFIG_NAME => Self::Avail(envy_load("da_avail_config", "DA_")?,
    Self::OBJECT_STORE_CLIENT_CONFIG_NAME => Self::ObjectStore(envy_load("da_object_stroe", DA_")?,
  };
  Ok(res)
}
```

Reply:
done